src/server/project.ts

```diff
@@ -211,52 +214,162 @@ namespace ts.server {
-                fileName => this.projectService.toPath(fileName),
-                () => this.compilerOptions,
-                directory => this.watchDirectoryOfFailedLookup(directory),
-                s => this.projectService.logger.info(s),
-                this.getProjectName(),
-                () => this.getTypeAcquisition().enable ? this.projectService.typingsInstaller.globalTypingsCacheLocation : undefined
-            );
-            this.lsHost.compilationSettings = this.compilerOptions;
-            this.resolutionCache.setModuleResolutionHost(this.lsHost);
-
-            this.languageService = createLanguageService(this.lsHost, this.documentRegistry);
+            if (host.realpath) {
+                this.realpath = path => host.realpath(path);
+            }
 
+            this.languageService = createLanguageService(this, this.documentRegistry);
             if (!languageServiceEnabled) {
                 this.disableLanguageService();
             }
-
+            this.resolutionCache = createResolutionCache(this);
             this.markAsDirty();
         }
 
-        private watchDirectoryOfFailedLookup(directory: string) {
+        getCompilationSettings() {
+            return this.compilerOptions;
+        }
+
+        getNewLine() {
+            return this.partialSystem.newLine;
+        }
+
+        getProjectVersion() {
+            return this.projectStateVersion.toString();
+        }
+
+        getScriptFileNames() {
+            const result: string[] = [];
+            if (this.rootFiles) {
+                this.rootFilesMap.forEach((value, _path) => {
+                    const f: ScriptInfo = isScriptInfo(value) && value;
+                    if (this.languageServiceEnabled || (f && f.isScriptOpen())) {
+                        // if language service is disabled - process only files that are open
+                        result.push(f ? f.fileName : value as NormalizedPath);
+                    }
+                });
+                if (this.typingFiles) {
+                    for (const f of this.typingFiles) {
+                        result.push(f);
+                    }
+                }
+            }
+            return result;
+        }
+
+        private getScriptInfoLSHost(fileName: string) {
+            const scriptInfo = this.projectService.getOrCreateScriptInfo(fileName, /*openedByClient*/ false, this.partialSystem);
+            if (scriptInfo) {
+                const existingValue = this.rootFilesMap.get(scriptInfo.path);
+                if (existingValue !== undefined && existingValue !== scriptInfo) {
+                    // This was missing path earlier but now the file exists. Update the root
+                    this.rootFiles.push(scriptInfo);
+                    this.rootFilesMap.set(scriptInfo.path, scriptInfo);
+                }
+                scriptInfo.attachToProject(this);
+            }
+            return scriptInfo;
+        }
+
+        getScriptKind(fileName: string) {

```

Comment:
Should we ever be calling this kind of function on a filename that doesn't have any associated script info?
Also, given that `getScriptInfoLSHost` has side effects, `getScriptKind` will too, is this intended?

Reply:
Yes i think it is intended. (these are the functions moved from lsHost to project